File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int result;
public:
    int maxPathSum(TreeNode* root) {
        result = root->val;
        pathSum(root);
        return result;
    }

    int pathSum(TreeNode* node){
        if (node == nullptr) return 0;

        int left = pathSum(node->left);
        int right = pathSum(node->right);
        int leftMax = max(left, 0);
        int rightMax = max(right, 0);

        result = max(result, node->val + leftMax + rightMax);
        return node->val + max(leftMax, rightMax);
    }
};

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int result;

    public int maxPathSum(TreeNode root) {
        this.result = root.val;
        pathSum(root);
        return result;
    }

    int pathSum(TreeNode node){
        if (node == null) return 0;

        int left = pathSum(node.left);
        int right = pathSum(node.right);
        int leftMax = Math.max(left, 0);
        int rightMax = Math.max(right, 0);

        this.result = Math.max(result, node.val + leftMax + rightMax);
        return node.val + Math.max(leftMax, rightMax);
    }
}